lib_eunix/lf_queue.ml

```diff
@@ -0,0 +1,64 @@
+(* A lock-free multi-producer, single-consumer, thread-safe queue without support for cancellation.
+   This makes a good data structure for a scheduler's run queue.
+
+   See: "Implementing lock-free queues"
+   https://people.cs.pitt.edu/~jacklange/teaching/cs2510-f12/papers/implementing_lock_free.pdf
+   
+   It is simplified slightly because we don't need multiple consumers.
+   Therefore [head] is not atomic. *)
+
+type 'a node = {
+  next : 'a node Atomic.t;
+  mutable value : 'a;
+}
+
+type 'a t = {
+  tail : 'a node Atomic.t;
+  mutable head : 'a node;
+}
+(* [head] is the last node dequeued (or a dummy node, initially).
+   [head.next] gives the real first node, if not [null].
+   [tail] is a node that was once at the tail of the queue.
+   Follow [tail.next] (if not [null]) to find the real last node. *)
+
+let null : 'a node = Obj.magic ()
+
+let atomic_null () : 'a node Atomic.t = Atomic.make (Obj.magic ())
+
+let push t x =
+  let node = { value = x; next = atomic_null () } in
+  let rec aux () =
```

Comment:
Are you sure this closure isn't allocated on every call?
Might be a long shot, but I have seen it happen. Often can be fixed by refactoring to:
 `let rec aux t node`
 where there are no captured variables in the closure.

Reply:
Good idea, but doesn't make any difference as far as I can see.